File: machine/cpp/machine/devices/vdp/fbm.cpp
#include "machine/devices/vdp/fbm.h"

namespace bmsx {

VdpFbmUnit::VdpFbmUnit(u32 width, u32 height) {
	configure(width, height);
}

void VdpFbmUnit::configure(u32 width, u32 height) {
	m_width = width;
	m_height = height;
	const size_t byteLength = static_cast<size_t>(width) * static_cast<size_t>(height) * 4u;
	m_displayFrameBufferCpuReadback.resize(byteLength);
	for (size_t index = 0u; index < byteLength; ++index) {
		m_displayFrameBufferCpuReadback[index] = 0u;
	}
}

std::vector<u8> VdpFbmUnit::captureDisplayReadback() const {
	std::vector<u8> pixels;
	pixels.resize(m_displayFrameBufferCpuReadback.size());
	for (size_t index = 0u; index < m_displayFrameBufferCpuReadback.size(); ++index) {
		pixels[index] = m_displayFrameBufferCpuReadback[index];
	}
	return pixels;
}

void VdpFbmUnit::restoreDisplayReadback(const std::vector<u8>& pixels) {
	m_displayFrameBufferCpuReadback.resize(pixels.size());
	for (size_t index = 0u; index < pixels.size(); ++index) {
		m_displayFrameBufferCpuReadback[index] = pixels[index];
	}
}

} // namespace bmsx

File: machine/ts/machine/devices/vdp/fbm.ts
export class VdpFbmUnit {
	private _width: number;
	private _height: number;
	private displayFrameBufferCpuReadback: Uint8Array;

	public constructor(width: number, height: number) {
		this._width = width;
		this._height = height;
		this.displayFrameBufferCpuReadback = new Uint8Array(width * height * 4);
	}

	public get width(): number {
		return this._width;
	}

	public get height(): number {
		return this._height;
	}

	public configure(width: number, height: number): void {
		this._width = width;
		this._height = height;
		this.displayFrameBufferCpuReadback = new Uint8Array(width * height * 4);
	}

	public captureDisplayReadback(): Uint8Array {
		const byteLength = this._width * this._height * 4;
		const pixels = new Uint8Array(byteLength);
		for (let index = 0; index < byteLength; index += 1) {
			pixels[index] = this.displayFrameBufferCpuReadback[index]!;
		}
		return pixels;
	}

	public restoreDisplayReadback(pixels: Uint8Array): void {
		for (let index = 0; index < pixels.byteLength; index += 1) {
			this.displayFrameBufferCpuReadback[index] = pixels[index]!;
		}
	}

	public get displayReadback(): Uint8Array {
		return this.displayFrameBufferCpuReadback;
	}
}
